# Drop Unneeded Columns and Transform Data to Numeric. 
## THIS IS THE LARGE SET.

The purpose of this notebook is to drop unneeded columns and to standardize data types to numeric. </br></br>
Outputs: </br>
**FEVS5year4.csv** contains numeric columns with no NaNs in the numeric values. There are NaN values only in the two string columns: 'agency_id' and 'StayorGo'. Use this CSV do feature anlaysis in Yellowbrick.</br>



In [1]:
#load libraries 
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
import pandas as pd
from scipy import stats
sns.set(style="ticks", color_codes=True)
%matplotlib inline
%config InlineBackend.figure_format = 'retina'

# 1. Read 2016 to 2020 FEVS survey data as Dataframes.

In [2]:
# need to give the path of these file on you own
df2016 = pd.read_csv('Standardized_Columns_fevs_2016_prdf.csv', low_memory=False)
df2017 = pd.read_csv('Standardized_Columns_fevs_2017_prdf.csv', low_memory=False)
df2018 = pd.read_csv('Standardized_Columns_fevs_2018_prdf.csv', low_memory=False)
df2019 = pd.read_csv('Standardized_Columns_fevs_2019_prdf.csv', low_memory=False)
df2020 = pd.read_csv('Standardized_Columns_fevs_2020_prdf.csv', low_memory=False)


In [3]:
#find cols' names in a dataframe: 
for col in df2020.columns:
    print(col)

response_id
year
random
agency_id
Q102
Q103
Q104
Q105
Q106
Q107
Q108
Q109
Q110
Q111
Q112
Q113
Q114
Q115
Q116
Q117
Q118
Q119
Q120
Q121
Q122
Q123
Q124
Q125
Q126
Q127
Q128
Q129
Q130
Q131
Q132
Q133
Q134
Q135
Q136
Q137
Q138
Q139
Q140
Q141
Q142
Q143
Q144
Q145
Q146
Q147
Q148
Q149
Q150
Q151
Q152
Q153
Q154
Q155
Q156
Q157
Q158
Q159
Q160
Q161
Q162
Q163
Q164
Q165
Q166
Q167
Q168
Q169
Q170
Q171
Q172
Q173
Q174
Q175
Q176
Q177
Q178
Q179
Q180
Q181
Q182
Q183
Q184
Q185
Q186
Q187
Q188
Q189
Q190
Q191
Q192
Q193
Q194
Q195
Q196
Q197
Q198
Q199
Q200
Q201
Q202
Q203
Q204
Q205
Q206
Q207
Q208
Q209
Q210
Q211
Q212
Q213
Q214
Q215
Q216
Q217
Q218
Q219
Q220
Q221
Q222
Q223
Q224
Q225
Q226
Q227
Q228
Q229
Q230
Q231
Q232
Q233
Q234
Q235
Q236
Q237
Q238
Q239
Q240
Q241
Q242
Q243
Q244
Q245
Q246
Q247
Q248
Q249
Q250
Q251
Q252
Q253
Q254
Q255
Q256
Q257
Q258
Q259
Q260
Q261
Q262
Q263
Q264
Q265
Q266
Q267
Q268
Q269
Q270
Q271
Q272
Q273
Q274
Q275
Q276
Q277
Q278
Q279
Q280
Q281
Q282
Q283
Q284
Q285
StayorGo


# 2. Combine 5 years of FEVS survey data into one dataframe.

call the new file FEVS5year

In [4]:
# combine 5 years of survey data into one datafram
frames = [df2016, df2017, df2018, df2019, df2020]
FEVS5year= pd.concat(frames)


In [5]:
FEVS5year.head()


response_id  year agency_id  Q234 Q226 Q228 Q225 Q236 Q237  Q102  ...  \
0   2016000002  2016        TR  TR93    A    B    B    A    A   5.0  ...   
1   2016000003  2016        AF  AF1C    A    A    B    B    A   NaN  ...   
2   2016000004  2016        TR  TRAD    A    A    B    B    A   5.0  ...   
3   2016000005  2016        TR  TR93    A    A    B    B    D   3.0  ...   
4   2016000006  2016        HE  HE09    B    B    B    B    A   5.0  ...   

   Q277  Q278  Q279  Q280  Q281  Q282  Q283 Q284 Q285 StayorGo  
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN  NaN  NaN     Stay  
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN  NaN  NaN     Stay  
2   NaN   NaN   NaN   NaN   NaN   NaN   NaN  NaN  NaN     Stay  
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN  NaN  NaN   Retire  
4   NaN   NaN   NaN   NaN   NaN   NaN   NaN  NaN  NaN     Stay  

[5 rows x 189 columns]

In [6]:
FEVS5year.tail()


response_id  year agency_id Q234 Q226 Q228 Q225 Q236 Q237  Q102  ...  \
624795   2020624797  2020        HS  NaN    A    B    B    A  NaN   5.0  ...   
624796   2020624798  2020        HS  NaN    A    A    B    A  NaN   5.0  ...   
624797   2020624799  2020        HS  NaN    A    B  NaN  NaN  NaN   4.0  ...   
624798   2020624800  2020        HS  NaN    A    B    A    B  NaN   4.0  ...   
624799   2020624801  2020        HS  NaN    B    B    A    B  NaN   4.0  ...   

        Q277  Q278  Q279  Q280  Q281  Q282  Q283 Q284 Q285  StayorGo  
624795   NaN   NaN   NaN   NaN   NaN   NaN   NaN  NaN  NaN  Transfer  
624796   NaN   NaN   NaN   NaN   NaN   NaN   NaN  NaN  NaN      Stay  
624797   NaN   NaN   NaN   NaN   NaN   NaN   NaN  NaN  NaN  Transfer  
624798   NaN   NaN   NaN   NaN   NaN   NaN   NaN  NaN  NaN    Retire  
624799   NaN   NaN   NaN   NaN   NaN   NaN   NaN  NaN  NaN      Stay  

[5 rows x 189 columns]

In [7]:
#check the number of numeric variables; expect 94
FEVS5year.describe()

response_id          year          Q102          Q238          Q103  \
count  2.732092e+06  2.732092e+06  2.716803e+06  2.094127e+06  2.699711e+06   
mean   2.018486e+09  2.018206e+03  3.710247e+00  3.756616e+00  3.626316e+00   
std    1.414725e+06  1.368376e+00  1.118916e+00  1.015602e+00  1.194728e+00   
min    2.016000e+09  2.016000e+03  1.000000e+00  1.000000e+00  1.000000e+00   
25%    2.017275e+09  2.017000e+03  3.000000e+00  3.000000e+00  3.000000e+00   
50%    2.018472e+09  2.018000e+03  4.000000e+00  4.000000e+00  4.000000e+00   
75%    2.019557e+09  2.019000e+03  4.000000e+00  4.000000e+00  5.000000e+00   
max    2.020625e+09  2.020000e+03  5.000000e+00  5.000000e+00  5.000000e+00   

               Q104          Q239          Q105          Q240          Q241  \
count  2.714021e+06  2.090605e+06  2.713057e+06  2.096759e+06  2.098584e+06   
mean   3.892961e+00  4.165207e+00  4.039856e+00  4.585513e+00  4.390788e+00   
std    1.075541e+00  9.005743e-01  9.545277e-01  6.487026e-01  7.205188e-01   
min    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    3.000000e+00  4.000000e+00  4.000000e+00  4.000000e+00  4.000000e+00   
50%    4.000000e+00  4.000000e+00  4.000000e+00  5.000000e+00  5.000000e+00   
75%    5.000000e+00  5.000000e+00  5.000000e+00  5.000000e+00  5.000000e+00   
max    5.000000e+00  5.000000e+00  5.000000e+00  5.000000e+00  5.000000e+00   

       ...           Q275           Q276           Q277           Q278  \
count  ...  324309.000000  324309.000000  324309.000000  324309.000000   
mean   ...       0.482854       0.213605       0.484149       0.690659   
std    ...       0.499707       0.409852       0.499749       0.462223   
min    ...       0.000000       0.000000       0.000000       0.000000   
25%    ...       0.000000       0.000000       0.000000       0.000000   
50%    ...       0.000000       0.000000       0.000000       1.000000   
75%    ...       1.000000       0.000000       1.000000       1.000000   
max    ...       1.000000       1.000000       1.000000       1.000000   

                Q279           Q280           Q281           Q282  \
count  324309.000000  324309.000000  324309.000000  324309.000000   
mean        0.303741       0.252528       0.449423       0.123768   
std         0.459873       0.434463       0.497436       0.329317   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         1.000000       1.000000       1.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

                Q283           Q284  
count  324309.000000  579912.000000  
mean        0.257742       3.616830  
std         0.437392       0.690771  
min         0.000000       1.000000  
25%         0.000000       3.000000  
50%         0.000000       4.000000  
75%         1.000000       4.000000  
max         1.000000       4.000000  

[8 rows x 94 columns]

### 94 out of 189 columns are numeric

In [8]:
# determine the number of NaN cases in Stay or Go
print(FEVS5year['StayorGo'].unique())
FEVS5year['StayorGo'].value_counts(dropna=False)

['Stay' 'Retire' 'Transfer' nan 'Leave']


Stay        1736122
Transfer     442466
Retire       278698
NaN          176148
Leave         98658
Name: StayorGo, dtype: int64

In [9]:
#total cases
all_records = 1736122+442466+278698+176148+98658
print("all_records =",all_records)

all_records = 2732092


In [10]:
#Percentage of NaN in StayorGo
NaN = 176148
print("NaN =",NaN / all_records)

NaN = 0.06447367072558317


In [11]:
#Percentage of Leave in StayorGo
Leave = 98658
print ("Leave =", Leave / all_records)

Leave = 0.036110789826989724


In [12]:
#Percentage of total Not_Stay in StayorGo
Stay = 1736122
Transfer = 442466
Retire = 278698
any_not_stay = 442466+278698+98658
print("Stay =", Stay / all_records)
print("Transfer =", Transfer /all_records)
print("Retire =", Retire / all_records)
print("Leave =", Leave / all_records)
print("NaN =", NaN / all_records)
print("--")
print("Stay =", Stay / all_records)
print("any_not_stay =", any_not_stay/all_records)
print("NaN =", NaN / all_records)


Stay = 0.6354551750087479
Transfer = 0.1619513544931869
Retire = 0.10200900994549232
Leave = 0.036110789826989724
NaN = 0.06447367072558317
--
Stay = 0.6354551750087479
any_not_stay = 0.30007115426566894
NaN = 0.06447367072558317


## 3. Drop 29 columns we do not need. Create Intermediate FEVS5year2:

In [13]:
FEVS5year2 = FEVS5year.drop(columns=['response_id','random','Q140','Q141','Q142','Q143','Q144','Q145','Q146','Q147','Q148','Q230','Q231','Q232','Q234','Q237','Q272','Q273','Q274','Q275','Q276','Q277','Q278','Q279','Q280','Q281','Q282','Q283','Q285'], errors='raise')
FEVS5year2


year agency_id Q226 Q228 Q225 Q236  Q102  Q238  Q103  Q104  ...  Q220  \
0       2016        TR    A    B    B    A   5.0   4.0   2.0   4.0  ...   NaN   
1       2016        AF    A    A    B    B   NaN   3.0   3.0   3.0  ...   NaN   
2       2016        TR    A    A    B    B   5.0   5.0   5.0   5.0  ...   NaN   
3       2016        TR    A    A    B    B   3.0   4.0   4.0   4.0  ...   NaN   
4       2016        HE    B    B    B    B   5.0   5.0   5.0   5.0  ...   NaN   
...      ...       ...  ...  ...  ...  ...   ...   ...   ...   ...  ...   ...   
624795  2020        HS    A    B    B    A   5.0   NaN   5.0   4.0  ...     X   
624796  2020        HS    A    A    B    A   5.0   NaN   5.0   4.0  ...     3   
624797  2020        HS    A    B  NaN  NaN   4.0   NaN   2.0   3.0  ...     1   
624798  2020        HS    A    B    A    B   4.0   NaN   3.0   4.0  ...     X   
624799  2020        HS    B    B    A    B   4.0   NaN   5.0   5.0  ...     1   

        Q221  Q222  Q223 Q224 Q227 Q229 Q235 Q284  StayorGo  
0        NaN   NaN   NaN  NaN  NaN  NaN  NaN  NaN      Stay  
1        NaN   NaN   NaN  NaN  NaN  NaN  NaN  NaN      Stay  
2        NaN   NaN   NaN  NaN  NaN  NaN  NaN  NaN      Stay  
3        NaN   NaN   NaN  NaN  NaN  NaN  NaN  NaN    Retire  
4        NaN   NaN   NaN  NaN  NaN  NaN  NaN  NaN      Stay  
...      ...   ...   ...  ...  ...  ...  ...  ...       ...  
624795     X     C     B    B    B    B  NaN  NaN  Transfer  
624796     4     A     B    B    B  NaN  NaN  NaN      Stay  
624797     1   NaN   NaN  NaN  NaN  NaN  NaN  NaN  Transfer  
624798     4     B     A    B    A    B  NaN  NaN    Retire  
624799     3     B     A    B    A    B  NaN  NaN      Stay  

[2732092 rows x 160 columns]

### 2.7M Rows  160 columns confirmed in FEVS5year2. How many are numeric?
Expect: 71

In [14]:
FEVS5year2.describe()

year          Q102          Q238          Q103          Q104  \
count  2.732092e+06  2.716803e+06  2.094127e+06  2.699711e+06  2.714021e+06   
mean   2.018206e+03  3.710247e+00  3.756616e+00  3.626316e+00  3.892961e+00   
std    1.368376e+00  1.118916e+00  1.015602e+00  1.194728e+00  1.075541e+00   
min    2.016000e+03  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    2.017000e+03  3.000000e+00  3.000000e+00  3.000000e+00  3.000000e+00   
50%    2.018000e+03  4.000000e+00  4.000000e+00  4.000000e+00  4.000000e+00   
75%    2.019000e+03  4.000000e+00  4.000000e+00  5.000000e+00  5.000000e+00   
max    2.020000e+03  5.000000e+00  5.000000e+00  5.000000e+00  5.000000e+00   

               Q239          Q105          Q240          Q241          Q110  \
count  2.090605e+06  2.713057e+06  2.096759e+06  2.098584e+06  2.722479e+06   
mean   4.165207e+00  4.039856e+00  4.585513e+00  4.390788e+00  4.015080e+00   
std    9.005743e-01  9.545277e-01  6.487026e-01  7.205188e-01  9.897640e-01   
min    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    4.000000e+00  4.000000e+00  4.000000e+00  4.000000e+00  4.000000e+00   
50%    4.000000e+00  4.000000e+00  5.000000e+00  5.000000e+00  4.000000e+00   
75%    5.000000e+00  5.000000e+00  5.000000e+00  5.000000e+00  5.000000e+00   
max    5.000000e+00  5.000000e+00  5.000000e+00  5.000000e+00  5.000000e+00   

       ...           Q211           Q212           Q213           Q214  \
count  ...  591907.000000  591907.000000  591907.000000  591907.000000   
mean   ...       0.113536       0.108948       0.130453       0.210161   
std    ...       0.317248       0.311574       0.336801       0.407423   
min    ...       0.000000       0.000000       0.000000       0.000000   
25%    ...       0.000000       0.000000       0.000000       0.000000   
50%    ...       0.000000       0.000000       0.000000       0.000000   
75%    ...       0.000000       0.000000       0.000000       0.000000   
max    ...       1.000000       1.000000       1.000000       1.000000   

                Q215           Q216           Q217           Q218  \
count  590013.000000  590013.000000  590013.000000  590013.000000   
mean        0.817123       0.085830       0.055382       0.040491   
std         0.386566       0.280114       0.228724       0.197107   
min         0.000000       0.000000       0.000000       0.000000   
25%         1.000000       0.000000       0.000000       0.000000   
50%         1.000000       0.000000       0.000000       0.000000   
75%         1.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

                Q219           Q284  
count  590013.000000  579912.000000  
mean        0.032767       3.616830  
std         0.178027       0.690771  
min         0.000000       1.000000  
25%         0.000000       3.000000  
50%         0.000000       4.000000  
75%         0.000000       4.000000  
max         1.000000       4.000000  

[8 rows x 71 columns]

### Answer: 71

## 4. Transform the demographic variables to categorical variables. Expect 174.

In [28]:
FEVS5year3 = pd.get_dummies(FEVS5year2, columns = ['Q222', 'Q223', 'Q224', 'Q225','Q226', 'Q227', 'Q228', 'Q229', 'Q235', 'Q236'])
print(FEVS5year3)

        year agency_id  Q102  Q238  Q103  Q104  Q239  Q105  Q240  Q241  ...  \
0       2016        TR   5.0   4.0   2.0   4.0   5.0   4.0   NaN   4.0  ...   
1       2016        AF   NaN   3.0   3.0   3.0   1.0   NaN   5.0   5.0  ...   
2       2016        TR   5.0   5.0   5.0   5.0   5.0   4.0   4.0   4.0  ...   
3       2016        TR   3.0   4.0   4.0   4.0   4.0   4.0   4.0   4.0  ...   
4       2016        HE   5.0   5.0   5.0   5.0   5.0   5.0   5.0   5.0  ...   
...      ...       ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   
624795  2020        HS   5.0   NaN   5.0   4.0   NaN   4.0   NaN   NaN  ...   
624796  2020        HS   5.0   NaN   5.0   4.0   NaN   4.0   NaN   NaN  ...   
624797  2020        HS   4.0   NaN   2.0   3.0   NaN   4.0   NaN   NaN  ...   
624798  2020        HS   4.0   NaN   3.0   4.0   NaN   4.0   NaN   NaN  ...   
624799  2020        HS   4.0   NaN   5.0   5.0   NaN   5.0   NaN   NaN  ...   

       Q227_C Q228_A Q228_B Q229_A Q229_B Q235_A Q2

### Confirmed: 160 columns + 24 new columns -10 original ABCD columns dropped = 174 columns

## View the updated dataframe. Expect 95 numeric columns.

In [29]:
FEVS5year3.describe()

year          Q102          Q238          Q103          Q104  \
count  2.732092e+06  2.716803e+06  2.094127e+06  2.699711e+06  2.714021e+06   
mean   2.018206e+03  3.710247e+00  3.756616e+00  3.626316e+00  3.892961e+00   
std    1.368376e+00  1.118916e+00  1.015602e+00  1.194728e+00  1.075541e+00   
min    2.016000e+03  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    2.017000e+03  3.000000e+00  3.000000e+00  3.000000e+00  3.000000e+00   
50%    2.018000e+03  4.000000e+00  4.000000e+00  4.000000e+00  4.000000e+00   
75%    2.019000e+03  4.000000e+00  4.000000e+00  5.000000e+00  5.000000e+00   
max    2.020000e+03  5.000000e+00  5.000000e+00  5.000000e+00  5.000000e+00   

               Q239          Q105          Q240          Q241          Q110  \
count  2.090605e+06  2.713057e+06  2.096759e+06  2.098584e+06  2.722479e+06   
mean   4.165207e+00  4.039856e+00  4.585513e+00  4.390788e+00  4.015080e+00   
std    9.005743e-01  9.545277e-01  6.487026e-01  7.205188e-01  9.897640e-01   
min    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    4.000000e+00  4.000000e+00  4.000000e+00  4.000000e+00  4.000000e+00   
50%    4.000000e+00  4.000000e+00  5.000000e+00  5.000000e+00  4.000000e+00   
75%    5.000000e+00  5.000000e+00  5.000000e+00  5.000000e+00  5.000000e+00   
max    5.000000e+00  5.000000e+00  5.000000e+00  5.000000e+00  5.000000e+00   

       ...        Q227_C        Q228_A        Q228_B        Q229_A  \
count  ...  2.732092e+06  2.732092e+06  2.732092e+06  2.732092e+06   
mean   ...  2.030869e-01  4.740832e-01  3.971360e-01  6.366037e-02   
std    ...  4.022967e-01  4.993280e-01  4.893047e-01  2.441470e-01   
min    ...  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    ...  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    ...  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
75%    ...  0.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
max    ...  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   

             Q229_B        Q235_A        Q235_B        Q235_C        Q236_A  \
count  2.732092e+06  2.732092e+06  2.732092e+06  2.732092e+06  2.732092e+06   
mean   1.505615e-01  1.510879e-01  1.842094e-01  1.950304e-01  2.733506e-01   
std    3.576210e-01  3.581346e-01  3.876549e-01  3.962242e-01  4.456794e-01   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
75%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  1.000000e+00   
max    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   

             Q236_B  
count  2.732092e+06  
mean   5.849591e-01  
std    4.927292e-01  
min    0.000000e+00  
25%    0.000000e+00  
50%    1.000000e+00  
75%    1.000000e+00  
max    1.000000e+00  

[8 rows x 95 columns]

### 95 columns are now numeric (net 14 more than previous of 71).
### Still need to transform 174-95=79 columns


In [30]:
#find cols' names in a dataframe: 
for col in FEVS5year3:
    print(col)
#columns are out of order

year
agency_id
Q102
Q238
Q103
Q104
Q239
Q105
Q240
Q241
Q242
Q106
Q107
Q108
Q243
Q244
Q245
Q246
Q109
Q247
Q248
Q110
Q249
Q250
Q111
Q113
Q251
Q252
Q253
Q254
Q114
Q255
Q115
Q256
Q257
Q258
Q116
Q259
Q260
Q261
Q117
Q118
Q119
Q120
Q262
Q263
Q121
Q264
Q122
Q123
Q124
Q265
Q125
Q126
Q127
Q128
Q266
Q129
Q267
Q130
Q268
Q131
Q132
Q133
Q134
Q135
Q136
Q269
Q270
Q271
Q137
Q138
Q139
Q233
Q112
Q149
Q150
Q151
Q152
Q153
Q154
Q155
Q156
Q157
Q158
Q159
Q160
Q161
Q162
Q163
Q164
Q165
Q166
Q167
Q168
Q169
Q170
Q171
Q172
Q173
Q174
Q175
Q176
Q177
Q178
Q179
Q180
Q181
Q182
Q183
Q184
Q185
Q186
Q187
Q188
Q189
Q190
Q191
Q192
Q193
Q194
Q195
Q196
Q197
Q198
Q199
Q200
Q201
Q202
Q203
Q204
Q205
Q206
Q207
Q208
Q209
Q210
Q211
Q212
Q213
Q214
Q215
Q216
Q217
Q218
Q219
Q220
Q221
Q284
StayorGo
Q222_A
Q222_B
Q222_C
Q222_D
Q223_A
Q223_B
Q224_A
Q224_B
Q225_A
Q225_B
Q226_A
Q226_B
Q227_A
Q227_B
Q227_C
Q228_A
Q228_B
Q229_A
Q229_B
Q235_A
Q235_B
Q235_C
Q236_A
Q236_B


In [31]:
#check a given variables for a question by uncommenting (command + '/') and running this cell
FEVS5year3.Q255.unique()


array(['4', '3', '5', '2', '1', nan, 'X'], dtype=object)

# 5. Coerce Numeric+X variables to Numbers only. 


### Identify the 77 columns where "Don't Know" was an option (appears as X in the data).
Coerce these variables to numbers. 

In [32]:
FEVS5year3[['Q106','Q107','Q108','Q109','Q111','Q112','Q113','Q114','Q115','Q116','Q117','Q119','Q120','Q121','Q122','Q127','Q128','Q129','Q130','Q131','Q132','Q133','Q163','Q164','Q165','Q166','Q167','Q168','Q182','Q183','Q184','Q185','Q186','Q187','Q188','Q189','Q190','Q191','Q192','Q193','Q194','Q195','Q196','Q197','Q198','Q205','Q206','Q207','Q208','Q209','Q220','Q221','Q242','Q243','Q244','Q245','Q246','Q247','Q248','Q249','Q250','Q251','Q252','Q253','Q255','Q256','Q257','Q258','Q259','Q260','Q261','Q262','Q263','Q264','Q266','Q267','Q268']]= FEVS5year3[['Q106','Q107','Q108','Q109','Q111','Q112','Q113','Q114','Q115','Q116','Q117','Q119','Q120','Q121','Q122','Q127','Q128','Q129','Q130','Q131','Q132','Q133','Q163','Q164','Q165','Q166','Q167','Q168','Q182','Q183','Q184','Q185','Q186','Q187','Q188','Q189','Q190','Q191','Q192','Q193','Q194','Q195','Q196','Q197','Q198','Q205','Q206','Q207','Q208','Q209','Q220','Q221','Q242','Q243','Q244','Q245','Q246','Q247','Q248','Q249','Q250','Q251','Q252','Q253','Q255','Q256','Q257','Q258','Q259','Q260','Q261','Q262','Q263','Q264','Q266','Q267','Q268']].apply(pd.to_numeric,errors='coerce')
FEVS5year3



year agency_id  Q102  Q238  Q103  Q104  Q239  Q105  Q240  Q241  ...  \
0       2016        TR   5.0   4.0   2.0   4.0   5.0   4.0   NaN   4.0  ...   
1       2016        AF   NaN   3.0   3.0   3.0   1.0   NaN   5.0   5.0  ...   
2       2016        TR   5.0   5.0   5.0   5.0   5.0   4.0   4.0   4.0  ...   
3       2016        TR   3.0   4.0   4.0   4.0   4.0   4.0   4.0   4.0  ...   
4       2016        HE   5.0   5.0   5.0   5.0   5.0   5.0   5.0   5.0  ...   
...      ...       ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   
624795  2020        HS   5.0   NaN   5.0   4.0   NaN   4.0   NaN   NaN  ...   
624796  2020        HS   5.0   NaN   5.0   4.0   NaN   4.0   NaN   NaN  ...   
624797  2020        HS   4.0   NaN   2.0   3.0   NaN   4.0   NaN   NaN  ...   
624798  2020        HS   4.0   NaN   3.0   4.0   NaN   4.0   NaN   NaN  ...   
624799  2020        HS   4.0   NaN   5.0   5.0   NaN   5.0   NaN   NaN  ...   

        Q227_C  Q228_A  Q228_B  Q229_A  Q229_B  Q235_A  Q235_B  Q235_C  \
0            0       0       1       0       0       0       0       0   
1            0       1       0       0       0       0       0       0   
2            0       1       0       0       0       0       0       0   
3            0       1       0       0       0       0       0       0   
4            0       0       1       0       0       0       0       0   
...        ...     ...     ...     ...     ...     ...     ...     ...   
624795       0       0       1       0       1       0       0       0   
624796       0       1       0       0       0       0       0       0   
624797       0       0       1       0       0       0       0       0   
624798       0       0       1       0       1       0       0       0   
624799       0       0       1       0       1       0       0       0   

        Q236_A  Q236_B  
0            1       0  
1            0       1  
2            0       1  
3            0       1  
4            0       1  
...        ...     ...  
624795       1       0  
624796       1       0  
624797       0       0  
624798       0       1  
624799       0       1  

[2732092 rows x 174 columns]

In [33]:
# Check that 172 variables are numeric (agency_id and StayorGo are remaining text variables)
FEVS5year3.describe()

year          Q102          Q238          Q103          Q104  \
count  2.732092e+06  2.716803e+06  2.094127e+06  2.699711e+06  2.714021e+06   
mean   2.018206e+03  3.710247e+00  3.756616e+00  3.626316e+00  3.892961e+00   
std    1.368376e+00  1.118916e+00  1.015602e+00  1.194728e+00  1.075541e+00   
min    2.016000e+03  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    2.017000e+03  3.000000e+00  3.000000e+00  3.000000e+00  3.000000e+00   
50%    2.018000e+03  4.000000e+00  4.000000e+00  4.000000e+00  4.000000e+00   
75%    2.019000e+03  4.000000e+00  4.000000e+00  5.000000e+00  5.000000e+00   
max    2.020000e+03  5.000000e+00  5.000000e+00  5.000000e+00  5.000000e+00   

               Q239          Q105          Q240          Q241          Q242  \
count  2.090605e+06  2.713057e+06  2.096759e+06  2.098584e+06  2.089196e+06   
mean   4.165207e+00  4.039856e+00  4.585513e+00  4.390788e+00  3.109714e+00   
std    9.005743e-01  9.545277e-01  6.487026e-01  7.205188e-01  1.277390e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    4.000000e+00  4.000000e+00  4.000000e+00  4.000000e+00  2.000000e+00   
50%    4.000000e+00  4.000000e+00  5.000000e+00  5.000000e+00  3.000000e+00   
75%    5.000000e+00  5.000000e+00  5.000000e+00  5.000000e+00  4.000000e+00   
max    5.000000e+00  5.000000e+00  5.000000e+00  5.000000e+00  5.000000e+00   

       ...        Q227_C        Q228_A        Q228_B        Q229_A  \
count  ...  2.732092e+06  2.732092e+06  2.732092e+06  2.732092e+06   
mean   ...  2.030869e-01  4.740832e-01  3.971360e-01  6.366037e-02   
std    ...  4.022967e-01  4.993280e-01  4.893047e-01  2.441470e-01   
min    ...  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    ...  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    ...  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
75%    ...  0.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
max    ...  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   

             Q229_B        Q235_A        Q235_B        Q235_C        Q236_A  \
count  2.732092e+06  2.732092e+06  2.732092e+06  2.732092e+06  2.732092e+06   
mean   1.505615e-01  1.510879e-01  1.842094e-01  1.950304e-01  2.733506e-01   
std    3.576210e-01  3.581346e-01  3.876549e-01  3.962242e-01  4.456794e-01   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
75%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  1.000000e+00   
max    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   

             Q236_B  
count  2.732092e+06  
mean   5.849591e-01  
std    4.927292e-01  
min    0.000000e+00  
25%    0.000000e+00  
50%    1.000000e+00  
75%    1.000000e+00  
max    1.000000e+00  

[8 rows x 172 columns]

In [34]:
#find cols' names in a dataframe: 
for col in FEVS5year3:
    print(col)
#columns are out of order

year
agency_id
Q102
Q238
Q103
Q104
Q239
Q105
Q240
Q241
Q242
Q106
Q107
Q108
Q243
Q244
Q245
Q246
Q109
Q247
Q248
Q110
Q249
Q250
Q111
Q113
Q251
Q252
Q253
Q254
Q114
Q255
Q115
Q256
Q257
Q258
Q116
Q259
Q260
Q261
Q117
Q118
Q119
Q120
Q262
Q263
Q121
Q264
Q122
Q123
Q124
Q265
Q125
Q126
Q127
Q128
Q266
Q129
Q267
Q130
Q268
Q131
Q132
Q133
Q134
Q135
Q136
Q269
Q270
Q271
Q137
Q138
Q139
Q233
Q112
Q149
Q150
Q151
Q152
Q153
Q154
Q155
Q156
Q157
Q158
Q159
Q160
Q161
Q162
Q163
Q164
Q165
Q166
Q167
Q168
Q169
Q170
Q171
Q172
Q173
Q174
Q175
Q176
Q177
Q178
Q179
Q180
Q181
Q182
Q183
Q184
Q185
Q186
Q187
Q188
Q189
Q190
Q191
Q192
Q193
Q194
Q195
Q196
Q197
Q198
Q199
Q200
Q201
Q202
Q203
Q204
Q205
Q206
Q207
Q208
Q209
Q210
Q211
Q212
Q213
Q214
Q215
Q216
Q217
Q218
Q219
Q220
Q221
Q284
StayorGo
Q222_A
Q222_B
Q222_C
Q222_D
Q223_A
Q223_B
Q224_A
Q224_B
Q225_A
Q225_B
Q226_A
Q226_B
Q227_A
Q227_B
Q227_C
Q228_A
Q228_B
Q229_A
Q229_B
Q235_A
Q235_B
Q235_C
Q236_A
Q236_B


### Success! 172 columns are numeric. 

In [35]:
#check a given variables for a question by uncommenting (command + '/') and running this cell
FEVS5year3.Q186.unique()
# FEVS5year3.agency_id.unique()
# FEVS5year3.StayorGo.unique()

array([nan,  5.,  3.,  2.,  4.,  1.])

In [36]:
#Or check all of them at once! 
#Note that the dtypes do not seem to be correct in the Output below.
# StayorGo appears as int64 but from the code above: array(['Stay', 'Retire', 'Transfer', nan, 'Leave'], dtype=object)
# Similarly, agency_id appears as int64 but from the code above: "array(['TR', 'AF', 'HE', 'CM', 'AG', 'GS', 'NN', 'DL', 'DJ', 'HU', 'HS'... dtype=object)"

col = ['agency_id','Q102','Q103','Q104','Q105','Q106','Q107','Q108','Q109','Q110','Q111','Q112','Q113','Q114','Q115','Q116','Q117','Q118','Q119','Q120','Q121','Q122','Q123','Q124','Q125','Q126','Q127','Q128','Q129','Q130','Q131','Q132','Q133','Q134','Q135','Q136','Q137','Q138','Q139','Q149','Q150','Q151','Q152','Q153','Q154','Q155','Q156','Q157','Q158','Q159','Q160','Q161','Q162','Q163','Q164','Q165','Q166','Q167','Q168','Q169','Q170','Q171','Q172','Q173','Q174','Q175','Q176','Q177','Q178','Q179','Q180','Q181','Q182','Q183','Q184','Q185','Q186','Q187','Q188','Q189','Q190','Q191','Q192','Q193','Q194','Q195','Q196','Q197','Q198','Q199','Q200','Q201','Q202','Q203','Q204','Q205','Q206','Q207','Q208','Q209','Q210','Q211','Q212','Q213','Q214','Q215','Q216','Q217','Q218','Q219','Q220','Q221','Q222_A','Q222_B','Q222_C','Q222_D','Q223_A','Q223_B','Q224_A','Q224_B','Q225_A','Q225_B','Q226_A','Q226_B','Q227_A','Q227_B','Q227_C','Q228_A','Q228_B','Q229_A','Q229_B','Q233','Q235_A','Q235_B','Q235_C','Q236_A','Q236_B','Q238','Q239','Q240','Q241','Q242','Q243','Q244','Q245','Q246','Q247','Q248','Q249','Q250','Q251','Q252','Q253','Q254','Q255','Q256','Q257','Q258','Q259','Q260','Q261','Q262','Q263','Q264','Q265','Q266','Q267','Q268','Q269','Q270','Q271','Q284','StayorGo','year']

#Print frequency of categories
for col in FEVS5year3:
    print ('\nFrequency of Categories for variable %s'%col)
    print (FEVS5year3[col].value_counts(dropna=False))


Frequency of Categories for variable year
2020    624800
2019    615395
2018    598003
2017    486105
2016    407789
Name: year, dtype: int64

Frequency of Categories for variable agency_id
HS    329891
AR    276735
HE    228556
TR    216010
AG    185111
NV    179736
AF    132660
IN    130184
DD    128249
DJ    120817
SZ    100310
TD     95918
VA     94707
CM     82146
NN     55961
EP     44009
DL     43310
DN     42757
ST     37424
GS     36197
XX     24960
HU     23324
OM     13571
SE     13355
ED     12819
NU     11242
AM     10316
NQ      8665
SB      7159
EE      6851
DR      5645
NF      4496
NL      4135
CU      3776
IB      3568
FT      2516
FC      2492
BG      2290
FQ      2158
RR      1528
SN      1509
CT      1395
BO      1367
HF       912
SK       712
KS       326
TC       317
Name: agency_id, dtype: int64

Frequency of Categories for variable Q102
4.0    1172777
5.0     678670
3.0     406779
2.0     316637
1.0     141940
NaN      15289
Name: Q102, dtype: int64

Frequency

5.0    1245130
4.0     780158
3.0     399398
2.0     141514
1.0     107209
NaN      58683
Name: Q126, dtype: int64

Frequency of Categories for variable Q127
4.0    834805
3.0    618113
2.0    431167
5.0    377975
1.0    339631
NaN    130401
Name: Q127, dtype: int64

Frequency of Categories for variable Q128
4.0    915033
3.0    570324
5.0    518954
1.0    256954
2.0    236862
NaN    233965
Name: Q128, dtype: int64

Frequency of Categories for variable Q266
4.0    887482
NaN    808752
5.0    459488
3.0    362051
2.0    111888
1.0    102431
Name: Q266, dtype: int64

Frequency of Categories for variable Q129
4.0    1207798
5.0     521416
3.0     468949
2.0     245643
1.0     171518
NaN     116768
Name: Q129, dtype: int64

Frequency of Categories for variable Q267
4.0    859154
NaN    830486
3.0    406386
5.0    369927
2.0    151631
1.0    114508
Name: Q267, dtype: int64

Frequency of Categories for variable Q130
4.0    1027885
3.0     519963
5.0     479438
2.0     317026
1.0     228405
N

NaN    2159017
5.0     275832
4.0     222407
3.0      58726
2.0      13010
1.0       3100
Name: Q190, dtype: int64

Frequency of Categories for variable Q191
NaN    2154907
5.0     335532
4.0     180217
3.0      46575
2.0      11626
1.0       3235
Name: Q191, dtype: int64

Frequency of Categories for variable Q192
NaN    2151520
5.0     332690
4.0     182054
3.0      51255
2.0      11656
1.0       2917
Name: Q192, dtype: int64

Frequency of Categories for variable Q193
NaN    2148912
5.0     339270
4.0     174685
3.0      52161
2.0      13386
1.0       3678
Name: Q193, dtype: int64

Frequency of Categories for variable Q194
NaN    2149070
5.0     308989
4.0     178241
3.0      68478
2.0      21716
1.0       5598
Name: Q194, dtype: int64

Frequency of Categories for variable Q195
NaN    2155212
5.0     295860
4.0     204627
3.0      59851
2.0      13113
1.0       3429
Name: Q195, dtype: int64

Frequency of Categories for variable Q196
NaN    2209535
5.0     225395
4.0     174043
3.0    

### We have 172 columns that are numeric. Result is that 147 variables still have NaN values.

## Save the FEVS5Year3 dataframe to a new CSV file before moving on.

In [37]:
# save a csv file to the local folder:
FEVS5year3.to_csv("FEVS5year3.csv", encoding='utf-8')

# FEVS5year3.csv has 172/174 numeric but still has NaN values in Number + X, Number (and agency_id, and StayorGo). 
#If we use this file for modeling, use Simple Imputer (see below) to transform the NaNs at the time of modeling.

# 6. Replace the resulting NaNs with mean for each column.


In [42]:
# #Replace NaNs in selected columns with the mean of values in the same column
#reimport the dataset
FEVS5year4 = pd.read_csv('FEVS5year3.csv', low_memory=False)


List of columns we want to impute.

In [43]:
# cols = FEVS5year4.columns[3:]

In [44]:
cols = ['Q102','Q238','Q103','Q104','Q239','Q105','Q240','Q241','Q242','Q106','Q107','Q108','Q243','Q244','Q245','Q246','Q109','Q247','Q248','Q110','Q249','Q250','Q111','Q113','Q251','Q252','Q253','Q254','Q114','Q255','Q115','Q256','Q257','Q258','Q116','Q259','Q260','Q261','Q117','Q118','Q119','Q120','Q262','Q263','Q121','Q264','Q122','Q123','Q124','Q265','Q125','Q126','Q127','Q128','Q266','Q129','Q267','Q130','Q268','Q131','Q132','Q133','Q134','Q135','Q136','Q269','Q270','Q271','Q137','Q138','Q139','Q233','Q112','Q149','Q150','Q151','Q152','Q153','Q154','Q155','Q156','Q157','Q158','Q159','Q160','Q161','Q162','Q163','Q164','Q165','Q166','Q167','Q168','Q169','Q170','Q171','Q172','Q173','Q174','Q175','Q176','Q177','Q178','Q179','Q180','Q181','Q182','Q183','Q184','Q185','Q186','Q187','Q188','Q189','Q190','Q191','Q192','Q193','Q194','Q195','Q196','Q197','Q198','Q199','Q200','Q201','Q202','Q203','Q204','Q205','Q206','Q207','Q208','Q209','Q210','Q211','Q212','Q213','Q214','Q215','Q216','Q217','Q218','Q219','Q220','Q221','Q284']

Create relationship between columns and the value you want to fill NA with.

In [45]:
impute = {}

for col in cols:
    impute[col] = FEVS5year4[col].mean()

In [46]:
impute

{'Q102': 3.7102465655404533,
 'Q238': 3.756615525228413,
 'Q103': 3.626316298300077,
 'Q104': 3.892961034568266,
 'Q239': 4.165207200786376,
 'Q105': 4.039856147511829,
 'Q240': 4.585512688868868,
 'Q241': 4.39078826484906,
 'Q242': 3.109714454747185,
 'Q106': 3.4032659999918766,
 'Q107': 3.5237915444560164,
 'Q108': 4.131096961589345,
 'Q243': 4.366517508298005,
 'Q244': 3.6939683390440456,
 'Q245': 3.8111283544141163,
 'Q246': 4.104742019408982,
 'Q109': 3.7264425946717683,
 'Q247': 3.431524499318247,
 'Q248': 3.8122537699342995,
 'Q110': 4.01508037343906,
 'Q249': 3.123208070378635,
 'Q250': 3.0321814588019214,
 'Q111': 2.934642049938087,
 'Q113': 3.1017778679743064,
 'Q251': 3.2033620770898126,
 'Q252': 3.897232516609701,
 'Q253': 3.564864158414267,
 'Q254': 4.272949546680637,
 'Q114': 3.9511253403477036,
 'Q255': 3.2496453886644354,
 'Q115': 3.4278533886783116,
 'Q256': 3.179909697020846,
 'Q257': 2.70011051352719,
 'Q258': 3.5948840636132635,
 'Q116': 3.944030011316906,
 'Q259': 

Apply fill NA.

In [47]:
#Replace NaNs in chosen columns with mean of that column
FEVS5year4.fillna(value=impute, inplace=True)

In [56]:
#check a given variables for a question by uncommenting (command + '/') and running this cell
FEVS5year4.Q102.unique()


array([5.        , 3.71024657, 3.        , 4.        , 2.        ,
       1.        ])

In [57]:
#Or check all of them at once! 
#Note that the dtypes do not seem to be correct in the Output below.
# StayorGo appears as int64 but from the code above: array(['Stay', 'Retire', 'Transfer', nan, 'Leave'], dtype=object)
# Similarly, agency_id appears as int64 but from the code above: "array(['TR', 'AF', 'HE', 'CM', 'AG', 'GS', 'NN', 'DL', 'DJ', 'HU', 'HS'... dtype=object)"

col = ['agency_id','Q102','Q103','Q104','Q105','Q106','Q107','Q108','Q109','Q110','Q111','Q112','Q113','Q114','Q115','Q116','Q117','Q118','Q119','Q120','Q121','Q122','Q123','Q124','Q125','Q126','Q127','Q128','Q129','Q130','Q131','Q132','Q133','Q134','Q135','Q136','Q137','Q138','Q139','Q149','Q150','Q151','Q152','Q153','Q154','Q155','Q156','Q157','Q158','Q159','Q160','Q161','Q162','Q163','Q164','Q165','Q166','Q167','Q168','Q169','Q170','Q171','Q172','Q173','Q174','Q175','Q176','Q177','Q178','Q179','Q180','Q181','Q182','Q183','Q184','Q185','Q186','Q187','Q188','Q189','Q190','Q191','Q192','Q193','Q194','Q195','Q196','Q197','Q198','Q199','Q200','Q201','Q202','Q203','Q204','Q205','Q206','Q207','Q208','Q209','Q210','Q211','Q212','Q213','Q214','Q215','Q216','Q217','Q218','Q219','Q220','Q221','Q222_A','Q222_B','Q222_C','Q222_D','Q223_A','Q223_B','Q224_A','Q224_B','Q225_A','Q225_B','Q226_A','Q226_B','Q227_A','Q227_B','Q227_C','Q228_A','Q228_B','Q229_A','Q229_B','Q233','Q235_A','Q235_B','Q235_C','Q236_A','Q236_B','Q238','Q239','Q240','Q241','Q242','Q243','Q244','Q245','Q246','Q247','Q248','Q249','Q250','Q251','Q252','Q253','Q254','Q255','Q256','Q257','Q258','Q259','Q260','Q261','Q262','Q263','Q264','Q265','Q266','Q267','Q268','Q269','Q270','Q271','Q284','StayorGo','year']

#Print frequency of categories
for col in FEVS5year4:
    print ('\nFrequency of Categories for variable %s'%col)
    print (FEVS5year4[col].value_counts(dropna=False))


Frequency of Categories for variable Unnamed: 0
0         5
271867    5
271865    5
271864    5
271863    5
         ..
618530    1
618531    1
618532    1
618533    1
624799    1
Name: Unnamed: 0, Length: 624800, dtype: int64

Frequency of Categories for variable year
2020    624800
2019    615395
2018    598003
2017    486105
2016    407789
Name: year, dtype: int64

Frequency of Categories for variable agency_id
HS    329891
AR    276735
HE    228556
TR    216010
AG    185111
NV    179736
AF    132660
IN    130184
DD    128249
DJ    120817
SZ    100310
TD     95918
VA     94707
CM     82146
NN     55961
EP     44009
DL     43310
DN     42757
ST     37424
GS     36197
XX     24960
HU     23324
OM     13571
SE     13355
ED     12819
NU     11242
AM     10316
NQ      8665
SB      7159
EE      6851
DR      5645
NF      4496
NL      4135
CU      3776
IB      3568
FT      2516
FC      2492
BG      2290
FQ      2158
RR      1528
SN      1509
CT      1395
BO      1367
HF       912
SK       

4.000000    740208
3.760431    690660
5.000000    631290
3.000000    353668
2.000000    182112
1.000000    134154
Name: Q264, dtype: int64

Frequency of Categories for variable Q122
4.00000    988755
5.00000    946597
3.00000    384466
2.00000    172209
1.00000    149701
3.91241     90364
Name: Q122, dtype: int64

Frequency of Categories for variable Q123
5.000000    1177779
4.000000     981507
3.000000     264852
2.000000     155984
1.000000      93453
4.119914      58517
Name: Q123, dtype: int64

Frequency of Categories for variable Q124
5.000000    1328766
4.000000     938137
3.000000     217782
2.000000     103429
1.000000      82597
4.245753      61381
Name: Q124, dtype: int64

Frequency of Categories for variable Q265
5.00000    852916
4.00000    845769
4.12554    682635
3.00000    166497
2.00000    123695
1.00000     60580
Name: Q265, dtype: int64

Frequency of Categories for variable Q125
5.000000    1136041
4.000000     820904
3.000000     364979
2.000000     187125
1.000000  

2.094468    2135645
3.000000     282164
1.000000     225819
2.000000      88464
Name: Q171, dtype: int64

Frequency of Categories for variable Q172
2.618699    2134035
3.000000     465178
1.000000      95161
2.000000      37718
Name: Q172, dtype: int64

Frequency of Categories for variable Q173
2.515527    2137192
3.000000     429483
1.000000     122796
2.000000      42621
Name: Q173, dtype: int64

Frequency of Categories for variable Q174
1.594428    2135547
1.000000     396587
3.000000     154645
2.000000      45313
Name: Q174, dtype: int64

Frequency of Categories for variable Q175
1.979465    2134866
1.000000     279211
3.000000     266947
2.000000      51068
Name: Q175, dtype: int64

Frequency of Categories for variable Q176
1.559062    2133344
1.000000     398390
3.000000     134379
2.000000      65979
Name: Q176, dtype: int64

Frequency of Categories for variable Q177
1.424233    2133206
1.000000     453707
3.000000     108888
2.000000      36291
Name: Q177, dtype: int64

Freque

0    1985273
1     746819
Name: Q236_A, dtype: int64

Frequency of Categories for variable Q236_B
1    1598162
0    1133930
Name: Q236_B, dtype: int64


In [58]:
FEVS5year4.head()

Unnamed: 0  year agency_id      Q102  Q238  Q103  Q104  Q239      Q105  \
0           0  2016        TR  5.000000   4.0   2.0   4.0   5.0  4.000000   
1           1  2016        AF  3.710247   3.0   3.0   3.0   1.0  4.039856   
2           2  2016        TR  5.000000   5.0   5.0   5.0   5.0  4.000000   
3           3  2016        TR  3.000000   4.0   4.0   4.0   4.0  4.000000   
4           4  2016        HE  5.000000   5.0   5.0   5.0   5.0  5.000000   

       Q240  ...  Q227_C  Q228_A  Q228_B  Q229_A  Q229_B  Q235_A  Q235_B  \
0  4.585513  ...       0       0       1       0       0       0       0   
1  5.000000  ...       0       1       0       0       0       0       0   
2  4.000000  ...       0       1       0       0       0       0       0   
3  4.000000  ...       0       1       0       0       0       0       0   
4  5.000000  ...       0       0       1       0       0       0       0   

   Q235_C  Q236_A  Q236_B  
0       0       1       0  
1       0       0       1  
2       0       0       1  
3       0       0       1  
4       0       0       1  

[5 rows x 175 columns]

In [61]:
FEVS5year4 = FEVS5year4.drop(columns=['Unnamed: 0'])

### Success! We have 173 numeric columns (172 + 1 "extra" column) with no NaN values in our dataframe of 175 columns.

somehow we added an "extra" Unnamed column at the start

## Save the FEVS5Year4 dataframe to a new CSV file.

In [51]:
# save out a csv file to the local folder:
FEVS5year4.to_csv("FEVS5year4.csv", encoding='utf-8')

